Shopify/hydrogen
# Automatically inject variables in runtime

### WHY are these changes introduced?

Working with environment variables has been a point of frustration for developers working with Hydrogen storefronts. This aims to streamline that experience and remove the cognitive load of understanding what environment variables are needed and where to put them.

### WHAT is this pull request doing?

Updates the `dev` command to automatically inject environment variables into the runtime instead of relying on the `envPath` argument.

<img width="566" alt="image" src="https://github.com/Shopify/hydrogen/assets/836758/03620c98-d75a-4afc-a568-b25507392019">

If a developer has a link to a Hydrogen storefront we'll automatically pull and inject their remote environment variables into the runtime. Developers can overwrite the remote variables by adding a different value to their `.env` file.

Special note: environment variables marked as secret cannot be injected because their values cannot be read through the API.

### HOW to test your changes?

1. Pull the branch and navigate to the `cli` directory
1. Run `npm run build`
1. Run `export SHOPIFY_HYDROGEN_FLAG_PATH=../../templates/demo-store` so we're always targeting the demo store
1. Run `h2 link` (note: if `h2` is unavailable run `npx shopify hydrogen shortcut` first) to see a list of your storefronts
   - You should be prompted to provide a shop
   - You should then be prompted to authenticate with that shop
   - You should be prompted to select a Hydrogen storefront from that shop
1. Remove all of the entires from the `.env` file in the demo store
1. Run `h2 dev` and note that your variables are automatically injected
1. Try overwriting one of the variables
   1. Stop the server
   1. Add an entry to the `.env` file in the demo store that uses the same key as your linked Hydrogen storefront
   1. Run `h2 dev` again
   1. Note that the variable is now overwritten by your local `.env`

#### Checklist
